### PR TITLE
App Plugins: support react pages and tabs

### LINKED ISSUE #16433: Migrate app pages to react
The existing angular pages let you build a set of pages and a config page.  The config page shows up at:
http://localhost:3000/plugins/ryantxu-firebase-app/edit

The app can include pages using somethign like:
```
  "includes": [
    {
      "type": "page",
      "name": "Debug",
      "component": "DebugPageCtrl",
      "role": "Viewer",
      "addToNav": true,
      "defaultNav": true
    },
```
That page then shows up under:
http://localhost:3000/plugins/ryantxu-firebase-app/page/debug

When we move to react, could the plugin expose a single component that handles all paths?  Essentially a root page with `<Router>`  This would expose everything under:
 *  http://localhost:3000/plugins/ryantxu-firebase-app/app/*
Or:
 *  http://localhost:3000/a/ryantxu-firebase-app/*
 *  http://localhost:3000/app/ryantxu-firebase-app/*

The `plugin.json` would be the same, but each Page would not need to map to a component.

This would allow apps to build much more rich UIs that support parameters.  

I think the react page should be in charge of the tabs/header, not the container.  See #11824  We can have a simple react component to match the standard header format

### PR DESCRIPTION
**Pull request by ryantxu**
fixes #16433

This issue has grown as I keep poking on it.  I'm not sure if I should split it up or just keep it together.  Let me know what makes sense.  

This PR includes:

* `AppRootPage`:  a single component that gets all request to /a/{plugin-id/*
* `PluginConfigTab`:  a way for apps (and eventually datasources, maybe panels) to add a tab on the existing plugin page
* Migrates the angular `PluginEditCtrl` to react `PluginPage`.  edit has always felt funny for everything except apps, since it just shows the readme.  I took "edit" out of the URL
* Loads the existing app `ConfigCtrl` within the react tabs -- seems to work!
* migrated angular DashImportListCtrl to react DashboardImporter.  Only used for apps right now
* improved the PluginMeta to support the properties used by these pages.  specifically dependencies and secureJsonData

I think it is close to ready to go, but would love someone else with apps to test it :)

Screenshots:
image

image

The app is configured with:
```
export const plugin = new AppPlugin()
  .setRootPage(ExampleRootPage)
  .addConfigTab({
    title: 'Tab 1',
    icon: 'fa fa-info',
    body: ExampleTab1,
    id: 'tab1',
  })
  .addConfigTab({
    title: 'Tab 2',
    icon: 'fa fa-user',
    body: ExampleTab2,
    id: 'tab2',
  });
```

TODO:
- [x] Open popup after clicking 'Update Available!' *blocker*
- [ ] Maybe someone with more react skilz should thunkify `PluginDashboards.tsx` and/or merge it with `DataSourceDashboards.tsx`
- [x] Once OK, remove 
   * [plugin_edit_ctrl.ts](https://github.com/grafana/grafana/blob/master/public/app/features/plugins/plugin_edit_ctrl.ts) 
   * [plugin_edit.ts](https://github.com/grafana/grafana/blob/master/public/app/features/plugins/partials/plugin_edit.html) 
   * [import_list](https://github.com/grafana/grafana/tree/master/public/app/features/plugins/import_list)

### CODE REVIEW

**Comment by ryantxu** [comment]
Consider: https://github.com/grafana/grafana/pull/16528#discussion_r275318407 when looking at the constructor

File: public/app/features/plugins/PluginPage.tsx [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+    alert('TODO, open modal');
+
+    const modalScope: any = {}; // this.$scope.$new(true);
+    modalScope.plugin = this.state.plugin.meta;
+
+    appEvents.emit('show-modal', {
+      src: 'public/app/features/plugins/partials/update_instructions.html',
+      scope: modalScope,
+    });
+  };
```
Comment (ryantxu) [inline public/app/features/plugins/PluginPage.tsx]: @torkelo -- this is my one real blocker, everything else just needs testing from someone else running apps :)

Any ideas?  is there a way to make a new $scope?  or to `show-modal` with a react component?  `update_instructions` would be pretty easy to convert

**Comment by ryantxu** [comment]
It would be great to get help testing this from someone running a bunch of plugins and apps.  This keeps still has both versions running so it is easy to compare:

New: React:
http://localhost:3333/plugins/example-app/

Old: Angular:
http://localhost:3333/plugins/example-app/edit

**Comment by ryantxu** [comment]
I think this is ready to go -- maybe someone with more react skilz should thunkify PluginDashboards.tsx and/or merge it with DataSourceDashboards.tsx, but I think that could happen later.

The part I think needs testing is the config page of existing apps.  It works fine for everything I have tried, but if apps got sneaky with the ConfigCtrl it may not work now.  

We should keep the concept of AppRootPage and ConfgTabs alpha for now, so I don't think this has anything we need to report externally yet.

**Review by torkelo — APPROVED** [review]
Looks good, great work! tested with Wordping & percona app & example app

**Comment by ryantxu** [comment]
@daniellee @briangann -- this PR changed the container for app ConfigCtrl elements to react.  Before 6.2, it would be good to check any apps you have on your systems to make sure they still work as expected :)

**Comment by Simplelenz** [comment]
Hi @ryantxu ,

How can I use this fix of plugins with tabs as navigation in my AngularJs plugins?  I recently joined a new company and they have a requirement like this but they have all their plugins built using AngularJs.

**Comment by torkelo** [comment]
we only support react plugins for this

**Pull request merged**